packages/core-components/src/providers/drawer/createDrawer.tsx

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  ComponentType,
+  DetailedHTMLProps,
+  HTMLAttributes,
+  PropsWithChildren,
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import { Drawer, DrawerProps } from '@material-ui/core';
+
+export interface DrawerComponentProps<T> {
+  value: T;
+  close: () => void;
+}
+
+export interface DrawerHook<T> {
+  openDrawer: (value: T) => void;
+  closeDrawer: () => void;
+}
+
+export type DrawerProviderProps = Omit<DrawerProps, 'open' | 'onClose'>;
+
+export interface ConstructedDrawer<T> {
+  Provider: ComponentType<PropsWithChildren<DrawerProviderProps>>;
+  useDrawer: () => DrawerHook<T>;
+}
+
+export type DrawerWrapperProps = DetailedHTMLProps<
+  HTMLAttributes<HTMLDivElement>,
+  HTMLDivElement
+>;
+
+export function createDrawer<T>(
+  Component: ComponentType<DrawerComponentProps<T>>,
+  wrapperProps?: DrawerWrapperProps,
+): ConstructedDrawer<T> {
+  const context = createContext<DrawerHook<T>>(undefined as any);
+
+  function Provider(props: PropsWithChildren<DrawerProviderProps>) {
+    const { children, ...drawerProps } = props;
+    const [openedValue, setOpenedValue] = useState<T | undefined>();
+
+    const closeDrawer = useCallback(
+      () => setOpenedValue(undefined),
+      [setOpenedValue],
+    );
+
+    const hookData = useMemo(
+      (): DrawerHook<T> => ({
+        openDrawer: setOpenedValue,
+        closeDrawer,
+      }),
+      [setOpenedValue, closeDrawer],
+    );
+
+    return (
+      <context.Provider value={hookData}>
+        {children}
+        <Drawer {...drawerProps} open={!!openedValue} onClose={closeDrawer}>
+          <div {...wrapperProps}>
+            {!openedValue ? null : (
+              <Component value={openedValue} close={closeDrawer} />
```

Comment:
This kinda makes me wonder whether we should pass the value through the `useDrawer` hook as well? That way we don't need these known props, the `Component` type gets a lot simpler, and we also avoid leaving a prop drilling problem for the `Component`.

Reply:
Hmm, which value?

The code that wants to _show_ the drawer calls `useDrawer().openDrawer()` essentially (and perhaps `.closeDrawer()`..

The component mounted _inside_ that drawer doesn't do `useDrawer()`. Or you think it should? 🤔 Yeah I guess that would work. Although... Should that component even know about the useDrawer hook? If so, you'd get a circular reference. The place where you create this drawer provider+hook needs to import the component to be "drawn". And that component needs to import the place the drawer provider+hook is created, to get the hook 😬

Perhaps a "global" `useCloseDrawer()` which can be imported from `@backstage/core-components` so that the inner component can close itself.

But the whole idea of that inner component to even know it's being inside a drawer is a bit odd. Maybe drawn-in components should be wrapped in a component which renders a close button, like an `x` at the top somewhere. For a uniform way of closing drawers (if clicking outside the drawer isn't obvious enough).